agent/app/agent_windows.go

```diff
@@ -0,0 +1,188 @@
+
+	// TODO: Pre-start hooks go here (unclear if we need any yet)
+
+	// https://msdn.microsoft.com/en-us/library/windows/desktop/ms682108(v=vs.85).aspx
+	// Not sure if a better link exists to describe what these values mean
+	accepts := svc.AcceptStop | svc.AcceptShutdown
+
+	// Announce that we are running and we accept the above-mentioned commands
+	responses <- svc.Status{State: svc.Running, Accepts: accepts}
+
+	defer func() {
+		// Announce that we are stopping
+		seelog.Info("Stopping Windows service")
+		responses <- svc.Status{State: svc.StopPending}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case r := <-requests:
+			switch r.Cmd {
+			case svc.Interrogate:
+				// Our status doesn't change unless we are told to stop or shutdown
+				responses <- r.CurrentStatus
+			case svc.Stop, svc.Shutdown:
+				return
+			default:
+				continue
+			}
+		}
+	}
+}
+
+// runAgent runs the ECS agent inside a goroutine and waits to be told to exit.
+func (h *handler) runAgent(ctx context.Context) {
+	agentCtx, cancel := context.WithCancel(ctx)
+	wg := sync.WaitGroup{}
+	running := true
+	terminationHandler := func(saver statemanager.Saver, taskEngine engine.TaskEngine) {
+		// We're using a waitgroup, a context, and a simple flag here.  The waitgroup gets added to as soon as this
+		// handler is invoked (agent.start() ultimately invokes it in a goroutine) so that at the end of the outer
+		// runAgent() function we know to wait for the handler to complete.  We then block on the context being
+		// canceled; this is our signal that the handler should actually run (happens either when the parent context is
+		// canceled because Windows told us to exit, or because the agent goroutine below exited unexpectedly).  The
+		// flag gets evaluated so that we know whether to actually save state; if the agent isn't properly running, we
+		// may not actually have any data to save.
+		wg.Add(1)
+		defer wg.Done()
+		<-agentCtx.Done()
+		if !running {
+			return
+		}
+
+		seelog.Info("Termination handler received signal to stop")
+		err := sighandlers.FinalSave(saver, taskEngine)
+		if err != nil {
+			seelog.Criticalf("Error saving state before final shutdown: %v", err)
+		}
+	}
+	h.ecsAgent.setTerminationHandler(terminationHandler)
+
+	go func() {
+		h.ecsAgent.start() // should block forever, unless there is an error
+		// TODO: distinguish between recoverable and unrecoverable errors
+		running = false
+		cancel()
+	}()
+
+	sleepCtx(agentCtx, time.Minute) // give the agent a minute to start and invoke terminationHandler
+
+	// wait for the termination handler to run.  Once the termination handler runs, we can safely exit.  If the agent
+	// exits by itself, the termination handler doesn't need to do anything and skips.  If the agent exits before the
+	// termination handler is invoked, we can exit immediately.
+	wg.Wait()
+}
+
+// sleepCtx provides a cancelable sleep
+func sleepCtx(ctx context.Context, duration time.Duration) {
```

Comment:
This can instead use the stdlib (or golang.net/x) [`context.WithDeadline`](https://godoc.org/golang.org/x/net/context#WithDeadline) function and wait on `ctx.Done()` of that context.